lib/functions.php

```diff
@@ -350,3 +354,76 @@ function put(string $path, string $contents): Promise
 {
     return filesystem()->put($path, $contents);
 }
+
+/**
+ * Asynchronously lock a file
+ * Resolves with a callbable that MUST eventually be called in order to release the lock.
+ *
+ * @param string            $file    File to lock
+ * @param bool              $shared  Whether to acquire a shared or exclusive lock (\LOCK_SH or \LOCK_EX, see PHP flock docs)
+ * @param integer           $polling Polling interval for lock in milliseconds
+ * @param CancellationToken $token   Cancellation token
+ *
+ * @return \Amp\Promise Resolves with a callbable that MUST eventually be called in order to release the lock.
+ */
+function lock(string $file, bool $shared, int $polling = 100, CancellationToken $token = null): Promise
+{
+    return call(static function () use ($file, $shared, $polling, $token) {
+        $operation = $shared ? \LOCK_SH : \LOCK_EX;
+        $token = $token ?? new NullCancellationToken;
+        if (!yield exists($file)) {
+            yield \touch($file);
+            StatCache::clear($file);
+        }
+        $operation |= LOCK_NB;
+        $res = \fopen($file, 'c');
+        do {
+            $result = \flock($res, $operation, $wouldblock);
+            if (!$result) {
+                if (!$wouldblock) {
+                    throw new FilesystemException("Failed acquiring lock on file.");
+                }
+                yield new Delayed($polling);
+                $token->throwIfRequested();
+            }
+        } while (!$result);
```

Comment:
Could be `while (!$result = \flock(...))` instead of `do` / `if` / `while`

Reply:
fixed